test/test_torch.py

```diff
@@ -5422,6 +5422,65 @@ def test_set_flush_denormal(self):
         self.assertEqual(double_tensor[2], 0.0, prec=0.0)  # tiny_double to zero
         torch.set_flush_denormal(False)
 
+    def test_unique(self):
+        x = torch.LongTensor([1, 2, 3, 2, 8, 5, 2, 3])
+        expected_unique = torch.LongTensor([1, 2, 3, 5, 8])
+        expected_inverse = torch.LongTensor([0, 1, 2, 1, 4, 3, 1, 2])
+        empty_inverse = torch.LongTensor([])
+
+        if TEST_NUMPY:
+            x_unique, x_inverse = torch.unique(x)
+            self.assertEqual(
+                expected_unique.numpy().tolist(),
+                sorted(x_unique.numpy().tolist())
+            )
+            self.assertEqual(empty_inverse, x_inverse)
+
+            x_unique, x_inverse = x.unique(return_inverse=True)
+            self.assertEqual(
+                expected_unique.numpy().tolist(),
+                sorted(x_unique.numpy().tolist())
+            )
+            self.assertEqual(expected_inverse.numel(), x_inverse.numel())
+
+        x_unique, x_inverse = x.unique(sorted=True)
+        self.assertEqual(expected_unique, x_unique)
+        self.assertEqual(empty_inverse, x_inverse)
+
+        x_unique, x_inverse = torch.autograd.Variable.unique(
+            x, sorted=True, return_inverse=True)
+        self.assertEqual(expected_unique, x_unique)
+        self.assertEqual(expected_inverse, x_inverse)
+
+        # Tests per-element unique on a higher rank tensor.
+        y = x.view(2, 2, 2)
+        y_unique, y_inverse = y.unique(sorted=True, return_inverse=True)
+        self.assertEqual(expected_unique, y_unique)
+        self.assertEqual(expected_inverse.view(y.size()), y_inverse)
+
+        # Tests unique on other types.
+        int_unique, int_inverse = torch.unique(
+            torch.IntTensor([2, 1, 2]), sorted=True, return_inverse=True)
+        self.assertEqual(torch.IntTensor([1, 2]), int_unique)
+        self.assertEqual(torch.LongTensor([1, 0, 1]), int_inverse)
+
+        double_unique, double_inverse = torch.unique(
+            torch.DoubleTensor([2., 1.5, 2.1, 2.]),
+            sorted=True,
+            return_inverse=True,
+        )
+        self.assertEqual(torch.DoubleTensor([1.5, 2., 2.1]), double_unique)
+        self.assertEqual(torch.LongTensor([1, 0, 2, 1]), double_inverse)
+
+        byte_unique, byte_inverse = torch.unique(
+            torch.ByteTensor([133, 7, 7, 7, 42, 128]),
+            sorted=True,
+            return_inverse=True,
+        )
+        self.assertEqual(torch.ByteTensor([7, 42, 128, 133]), byte_unique)
+        self.assertEqual(torch.LongTensor([3, 0, 0, 0, 1, 2]), byte_inverse)
```

Comment:
It would be better to implement `unique` reference implementation in Python, and then have a list of examples that we loop over and compare outputs. Here's an example:
```python
def unique(tensor):
    values = sorted(set(tensor))
    value_map = {i: v for i, v in enumerate(values)}
    inverse = torch.LongTensor([value_map[v.item()] for v in values])
    return values, inverse
```

Reply:
Hmm - for unit tests, I usually prefer to err on the side of verbosity and explicitness rather than doing things programmatically - since that makes it easier for people to see what the expected return values are and also serve as additional documentation for what we expect the function to return.

I could change to doing a reference implementation though, if you feel strongly about this.